### PR TITLE
feat: add fuzzy search to command palette

### LINKED ISSUE #758: type(ui): implement true fuzzy search for Command Palette
### Objective

Implement true fuzzy search and result ranking for the CommandPalette widget so that non-consecutive character queries can match commands intuitively.

### Package scope

packages/widgets

### Files to create or modify

MODIFY: packages/widgets/src/input/CommandPalette.ts

MODIFY: packages/widgets/src/input/CommandPalette.test.ts

### API contract

```typescript
No public API changes required.

The existing CommandPalette API should remain fully backward compatible.

The filtering implementation should be enhanced internally to:

Support fuzzy matching on label
Support fuzzy matching on id
Rank stronger matches above weaker matches
Preserve existing case-insensitive behavior
```

### Acceptance criteria

Query open matches Open File

Query opf matches Open File

Query stng matches Settings

Query cmp matches Command Palette

Exact matches rank above fuzzy matches

Search remains case-insensitive

Existing substring searches continue to work

Selection index remains valid after filtering

All existing CommandPalette tests pass

bun vitest run packages/widgets passes

bun run typecheck passes

### Test expectations

File: packages/widgets/src/input/CommandPalette.test.ts

Run:

bun vitest run packages/widgets

Cases:

exact match returns expected command
fuzzy match works with non-consecutive characters
exact match ranks above fuzzy match
matching is case-insensitive
command id participates in fuzzy matching
selection index is clamped after filtering
existing substring behavior remains functional

### Reference pattern to follow

Follow:

packages/widgets/src/input/CommandPalette.ts

Follow test structure from:

packages/widgets/src/data/Gauge.test.ts

### Not included (out of scope)

No new widget creation
No external fuzzy-search libraries
No dependency additions
No JSX changes
No docs-site changes
No UI redesign

### Do not touch

Any package other than packages/widgets
packages/core
bun.lock
.github/

### Agent prompt (copy this into your AI agent)

```text
Read AGENTS.md at the repo root and packages/widgets/AGENTS.md.

Read packages/widgets/src/input/CommandPalette.ts and understand the current substring filtering implementation.

Task: replace the current substring-only filtering logic with a lightweight fuzzy matching implementation while keeping the public API unchanged.

Implement match scoring and ranking so that stronger matches appear first.

Add comprehensive tests in CommandPalette.test.ts covering exact matches, fuzzy matches, ranking behavior, case-insensitive matching, and id matching.

Plan first. List files to change and confirm they match this issue.

Do not add dependencies.

When finished run:

bun vitest run packages/widgets && bun run typecheck
```

### Difficulty

advanced

### Before you open the PR

- [x] I run bun run build && bun vitest run && bun run typecheck and all pass
- [x] My change is confined to the package scope above
- [x] bun.lock has no unrelated changes
- [x] I starred the repo

### PR DESCRIPTION
**Pull request by Ishika-1110**
## Description

This PR adds fuzzy search support to the `CommandPalette` widget in `@termuijs/widgets`.

### Changes made

* Added a fuzzy matching helper for command filtering.
* Updated the filtering logic to support non-contiguous character matches.
* Preserved existing case-insensitive search behavior.
* Added comprehensive test coverage for fuzzy search functionality and edge cases.

## Related Issue

Closes #758

## Which package(s)?

@termuijs/widgets

## Type of Change

* [ ] 🐛 Bug fix (`type:bug`)
* [x] ✨ Feature (`type:feature`)
* [ ] 📝 Docs (`type:docs`)
* [x] 🧪 Tests (`type:testing`)
* [ ] ♻️ Refactor (`type:refactor`)
* [ ] 🎨 Design / UX (`type:design`)
* [ ] ♿ Accessibility (`type:accessibility`)
* [ ] ⚡ Performance (`type:performance`)
* [ ] 🔧 DevOps / CI (`type:devops`)
* [ ] 🔒 Security (`type:security`)

## Checklist

* [x] ⭐ You starred the repo. The `needs-star` check blocks your merge otherwise.
* [x] Tests pass locally: `bun vitest run`
* [x] Build passes: `bun run build`
* [x] Typecheck passes: `bun run typecheck`
* [x] You read `CONTRIBUTING.md`.
* [x] Your PR title follows `type: short description`.
* [x] Widget state mutators call `markDirty()` (if your change affects rendering).
* [x] No new `any` types without an inline comment explaining why.
* [x] No unrelated refactors bundled into this PR.

## GSSoC 2026 Participation

* [x] You are a GSSoC 2026 contributor.
* Your GSSoC profile: https://gssoc.girlscript.org/profile/84e87b15-54d6-40b1-949b-a09fec98075a

## Screenshots / Recordings (UI changes)

N/A – functionality and test updates only.

## Notes for the Reviewer

* Added fuzzy matching support to improve command discoverability.
* Existing substring search behavior remains supported.
* Added tests covering:

  * Non-contiguous fuzzy matches
  * Case-insensitive matching
  * No-match scenarios
* All existing CommandPalette tests continue to pass.

## Summary by CodeRabbit

* **New Features**
  * Command palette now uses ranked fuzzy search for more flexible discovery: supports non-contiguous and case-insensitive matching and ranks exact > substring > fuzzy so the best match is selected.

* **Tests**
  * Added test coverage validating fuzzy matching, case-insensitivity, ranking, and selection behavior.

### CODE REVIEW

File: packages/widgets/src/input/CommandPalette.test.ts [hunk 1]
```diff
@@ -299,4 +299,41 @@ describe('CommandPalette', () => {
...
+
+            'xyz'.split('').forEach((ch) => palette.handleKey(ch));
+
+            palette.handleKey('enter');
+
+            cmds.forEach((cmd) =>
+                expect(cmd.action).not.toHaveBeenCalled()
+            );
+        });
+    });
```
[the remark this answers is not in the document]
Comment (Ishika-1110) [reply packages/widgets/src/input/CommandPalette.test.ts:396]: Added ranking tests for exact, substring, and fuzzy match ordering.
Updated fuzzy search comments to reflect current implementation.

**Comment by Karanjot786** [comment]
Hi @Ishika-1110,

This PR is approved for merge.

The fuzzy search addition to CommandPalette is well-done. Supporting non-contiguous character matches and sorting by exact then substring then fuzzy score covers the important cases. The `_fuzzyScore` helper is a clean private method.

Two minor style things worth cleaning up:
1. The two new `it()` blocks at the bottom of the test file appear outside the `describe` block visually, though they compile correctly. Move them inside the `describe` block for clarity.
2. The comment in `_filter` still says "case-insensitive substring" but it now does fuzzy matching. Update it.

Neither blocks merge. Thank you for the contribution.
